manage.py

```diff
@@ -1,21 +1,153 @@
-def main():
+
+DEFAULT_ENVS = {
+    "DJANGO_SETTINGS_MODULE": "pydis_site.settings",
+    "SUPER_USERNAME": "admin",
+    "SUPER_PASSWORD": "admin",
+    "DEFAULT_BOT_API_KEY": "badbot13m0n8f570f942013fc818f234916ca531",
+}
+
+
+for key, value in DEFAULT_ENVS.items():
+    os.environ.setdefault(key, value)
+
+
+class SiteManager:
+    """
+    Manages the preparation and serving of the website.
+
+    Handles both development and production environments.
+
+    Usage:
+        manage.py run [option]...
+
+    Options:
+        --debug    Runs a development server with debug mode enabled.
+        --silent   Sets minimal console output.
+        --verbose  Sets verbose console output.
+    """
+
+    def __init__(self, args: List[str]):
+        self.debug = "--debug" in args
+        self.silent = "--silent" in args
+
+        if self.silent:
+            self.verbosity = 0
+        else:
+            self.verbosity = 2 if "--verbose" in args else 1
+
+        if self.verbosity and self.debug:
+            os.environ.setdefault("DEBUG", "true")
+            print("Starting in debug mode.")
+
+    @staticmethod
+    def create_superuser() -> None:
+        """Create a default django admin super user in development environments."""
+        print("Creating a superuser.")
+
+        name = os.environ["SUPER_USERNAME"]
+        password = os.environ["SUPER_PASSWORD"]
+        bot_token = os.environ["DEFAULT_BOT_API_KEY"]
+        user = get_user_model()
+
+        # Get or create admin superuser.
+        if user.objects.filter(username=name).exists():
+            user = user.objects.get(username=name)
+            print('Admin superuser already exists.')
+        else:
+            user = user.objects.create_superuser(name, '', password)
+            print('Admin superuser created.')
+
+        # Setup a default bot token to connect with site API
+        from rest_framework.authtoken.models import Token
+        token, is_new = Token.objects.update_or_create(user=user)
+        if token.key != bot_token:
+            token.delete()
+        token, is_new = Token.objects.update_or_create(user=user, key=bot_token)
+        if is_new:
+            print(f"New bot token created: {token}")
+        else:
+            print(f"Existing bot token found: {token}")
+
+    @staticmethod
+    def wait_for_postgres() -> None:
+        """Wait for the PostgreSQL database specified in DATABASE_URL."""
+        print("Waiting for PostgreSQL database.")
+
+        # Get database URL based on environmental variable passed in compose
+        database_url = os.environ["DATABASE_URL"]
+        match = re.search(r"@(\w+):(\d+)/", database_url)
+        if not match:
+            raise OSError("Valid DATABASE_URL environmental variable not found.")
+        domain = match.group(1)
+        port = int(match.group(2))
+
+        # Attempt to connect to the database socket
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        while True:
```

Comment:
What would happen if the container crashed? Would this just perpetually wait or would it raise an exception that isn't `socket.error`?

Reply:
If the PGSQL container failed to start, it would keep waiting and outputting `retrying` as the socket only doesn't error when the port is open, meaning the PGSQL container is fully ready to accept connections.

This also is only done during dev environments, so it's a good reminder that they may have forgotten to up the other container (dunno why they're running `docker-compose site` in that case though), or they severely broke something.

Reply:
Do you think it would be good to add a timeout and exit with return code 1?

Reply:
Hmm, I'm not sure, as it's pretty clear what it's doing and it's simple to interrupt. I guess though if they run the container detached and it has that issue then it would just keep going and going I guess until they realise by manually checking logs, so for just that possible edge case, it might make sense.

Reply:
On testing, I realised it's actually impossible to launch web alone via compose thanks to `depends_on` anyway, so the only case is if something is really really broken somehow. At any rate, I've added and tested the retry maximum.